File: backend/app/schemas/job_board.py
import uuid
from datetime import date, datetime
from typing import Literal, Optional

from pydantic import BaseModel

KindLiteral = Literal["ofrece", "busca"]


class JobPostCreate(BaseModel):
    kind: KindLiteral
    title: str
    description: str
    #: Cómo contactar al autor. Lo escribe él: puede no querer dar su teléfono.
    contact: str
    category: Optional[str] = None


class JobPostUpdate(BaseModel):
    kind: Optional[KindLiteral] = None
    title: Optional[str] = None
    description: Optional[str] = None
    contact: Optional[str] = None
    category: Optional[str] = None


class JobPostModeration(BaseModel):
    approve: bool
    #: Motivo del rechazo. Sin él, el autor vuelve a mandar lo mismo.
    note: Optional[str] = None
    #: Días de vigencia al aprobar.
    days: Optional[int] = None


class JobAttachmentResponse(BaseModel):
    id: uuid.UUID
    url: str
    filename: str
    content_type: str
    size_bytes: int
    is_image: bool


class JobPostResponse(BaseModel):
    id: uuid.UUID
    kind: str
    title: str
    description: str
    #: Arranque del texto, ya sin marcas de formato: es lo que se lee en la
    #: tarjeta del listado. Se calcula en el servidor para que la tarjeta y la
    #: página no puedan recortar distinto.
    excerpt: str = ""
    cover_image_url: Optional[str] = None
    attachments: list[JobAttachmentResponse] = []
    #: `None` en avisos que no están vigentes y no son propios: un teléfono de un
    #: socio no circula sin motivo.
    contact: Optional[str] = None
    category: Optional[str] = None
    #: `vencido` es derivado de la fecha, no un estado guardado.
    status: str
    #: Sólo lo ve el autor: es la explicación de por qué se rechazó.
    moderation_note: Optional[str] = None
    author_name: str
    #: Iniciales del autor, para el avatar de la tarjeta. Va resuelto del servidor
    #: porque partir un nombre en el cliente termina distinto en cada pantalla.
    author_initials: str = ""
    is_mine: bool
    published_at: Optional[datetime] = None
    expires_on: Optional[date] = None

File: frontend/src/lib/jobBoard.ts
/**
 * Tipos y etiquetas de la bolsa, compartidos por el feed y la página del aviso.
 *
 * Estaban dentro de la pantalla, que era la única. Con dos, duplicarlos significa
 * que un día el estado dice "vencido" en una y "expirado" en la otra.
 */

export interface JobAttachment {
  id: string;
  url: string;
  filename: string;
  content_type: string;
  size_bytes: number;
  is_image: boolean;
}

export interface JobPost {
  id: string;
  kind: "ofrece" | "busca";
  title: string;
  description: string;
  /** Arranque del texto sin marcas de formato: lo que se lee en la tarjeta. */
  excerpt: string;
  cover_image_url: string | null;
  attachments: JobAttachment[];
  contact: string | null;
  category: string | null;
  status: "pendiente" | "publicado" | "rechazado" | "vencido";
  moderation_note: string | null;
  author_name: string;
  author_initials: string;
  is_mine: boolean;
  published_at: string | null;
  expires_on: string | null;
}

export const KIND_LABEL: Record<string, string> = {
  ofrece: "Ofrece trabajo",
  busca: "Busca trabajo",
};

export const STATUS_CLASS: Record<string, string> = {
  pendiente: "bg-amber-100 text-amber-700",
  publicado: "bg-brand-soft text-brand",
  rechazado: "bg-red-100 text-red-700",
  vencido: "bg-surface-strong text-ink-muted",
};

export function daysLeft(expires: string | null): string | null {
  if (!expires) return null;
  const diff = Math.ceil((new Date(expires).getTime() - Date.now()) / 86_400_000);
  if (diff < 0) return "vencido";
  if (diff === 0) return "vence hoy";
  if (diff === 1) return "vence mañana";
  return `vence en ${diff} días`;
}

export function formatSize(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${Math.round(bytes / 1024)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}
